Geometry/DTGeometryBuilder/src/DTGeometryParsFromDD.cc

```diff
@@ -232,13 +256,153 @@ DTGeometryParsFromDD::PosRotPair DTGeometryParsFromDD::plane(const DDFilteredVie
   grmat[7] = z.Y();
   grmat[8] = z.Z();
 
-  //   std::cout << "rotation by its own operator: "<< tmp << std::endl;
-  //   DD3Vector tx, ty,tz;
-  //   tmp.GetComponents(tx, ty, tz);
-  //   std::cout << "rotation manually: "
-  // 	    << tx.X() << ", " << tx.Y() << ", " << tx.Z() << std::endl
-  // 	    << ty.X() << ", " << ty.Y() << ", " << ty.Z() << std::endl
-  // 	    << tz.X() << ", " << tz.Y() << ", " << tz.Z() << std::endl;
+  return pair<std::vector<double>, std::vector<double> >(gtran, grmat);
+}
+
+// DD4Hep
+
+void DTGeometryParsFromDD::buildGeometry(cms::DDFilteredView& fv,
+                                         const MuonGeometryConstants& muonConstants,
+                                         RecoIdealGeometry& rig) const {
+  edm::LogVerbatim("DTGeometryParsFromDD") << "(0) DTGeometryParsFromDD - DD4HEP ";
+
+  bool doChamber = fv.firstChild();
+
+  while (doChamber) {
+    insertChamber(fv, muonConstants, rig);
+
+    bool doSL = fv.nextSibling();
+    while (doSL) {
+      insertSuperLayer(fv, muonConstants, rig);
+
+      fv.down();
+      bool doLayers = fv.sibling();
+      while (doLayers) {
+        insertLayer(fv, muonConstants, rig);
+
+        doLayers = fv.sibling();
+      }
+
+      doSL = fv.nextSibling();
+    }
+
+    fv.parent();
+    doChamber = fv.firstChild();
+  }
+}
+
+DTGeometryParsFromDD::PosRotPair DTGeometryParsFromDD::plane(const cms::DDFilteredView& fv) const {
+  const Double_t* tr = fv.trans();
+
+  std::vector<double> gtran(3);
+
+  gtran[0] = tr[0] / dd4hep::cm;
+  gtran[1] = tr[1] / dd4hep::cm;
+  gtran[2] = tr[2] / dd4hep::cm;
+
+  edm::LogVerbatim("DTGeometryParsFromDD") << "(4) DD4HEP, Position "
+                                           << " " << gtran[0] << " " << gtran[1] << " " << gtran[2];
+
+  DDRotationMatrix rotation = fv.rotation();
+  DD3Vector x, y, z;
+  rotation.GetComponents(x, y, z);
+
+  std::vector<double> grmat(9);
+
+  grmat[0] = x.X();
+  grmat[1] = x.Y();
+  grmat[2] = x.Z();
+
+  grmat[3] = y.X();
+  grmat[4] = y.Y();
+  grmat[5] = y.Z();
+
+  grmat[6] = z.X();
+  grmat[7] = z.Y();
+  grmat[8] = z.Z();
 
   return pair<std::vector<double>, std::vector<double> >(gtran, grmat);
 }
+
+void DTGeometryParsFromDD::insertChamber(cms::DDFilteredView& fv,
+                                         const MuonGeometryConstants& muonConstants,
+                                         RecoIdealGeometry& rig) const {
+  MuonGeometryNumbering mdddnum(muonConstants);
+  DTNumberingScheme dtnum(muonConstants);
+  int rawid = dtnum.baseNumberToUnitNumber(mdddnum.geoHistoryToBaseNumber(fv.history()));
+  DTChamberId detId(rawid);
+
+  vector<double> par_temp = fv.parameters();
+  vector<double> par(4);
+  par[0] = 0;  // 0 is the ID of a Chamber, please see DTGeometryBuilderCondDB.cc
```

Comment:
```suggestion
  par[0] = DTChamberTag;  // DTChamberTag is the ID of a Chamber.
```

Reply:
See lines 137, 163, and 185 of this file. There are enum constants defined as the IDs of each kind of component. It would be good to change `DTGeometryBuilderCondDB.cc` to use these tags instead of 0, 1, and 2.

Reply:
Done